### PR TITLE
Spawn a process for printing

### PR DESCRIPTION
**Pull request by skalnik**
This uses Process.spawn to run `make`, and waits for it for 5 seconds. If the process doesn't exit by then, we assume the print is going OK :+1:

/cc @sshirokov

### CODE REVIEW

**Comment by skalnik** [comment]
Added a `/log` endpoint so we can see the log being printed as we wait for prints.

File: server/app.rb [hunk 1]
```diff
@@ -42,24 +43,36 @@ class App < Sinatra::Base
...
+                   "#{File.dirname(stl_file)}/#{File.basename(stl_file, '.stl')}",
+                   "| tee #{LOG_FILE}"].join " "
+
       begin
-        PrintMe::Download.new(stl_url, 'data/print.stl').fetch
-        if system('make data/print')
-          status 201
-          "Thing printed! Go pick it up"
-        else
+        Timeout::timeout(5) do
```
Comment (sshirokov) [inline server/app.rb:59]: Can you move the timeout to live wrap just the `Process.wait` just to communicate intent?

File: server/app.rb [hunk 2]
```diff
@@ -42,24 +43,36 @@ class App < Sinatra::Base
...
+          pid = Process.spawn(make_stl)
```
Comment (sshirokov) [inline server/app.rb]: Instead of the `|tee` args, why not use :out and :err for `Process#spawn`?

File: server/app.rb (hunk 1 above)
Comment (skalnik) [inline server/app.rb:59]: Good idea. I also realized we should clear out the pid file if it dies that quickly.

File: server/app.rb (hunk 2 above)
Comment (skalnik) [inline server/app.rb]: I didn't even know it had 'em!

**Comment by skalnik** [comment]
Any other thoughts @sshirokov?

**Comment by sshirokov** [comment]
Let's not kill th pidfile until an unlock. It seems useful 'is it running' info

**Comment by skalnik** [comment]
What's the point of the lockfile then? I think the pidfile should only be for the actual `make` command, since we also wanna add a way of canceling a print by killing the process, and that would look up the pid and then kill the pidfile too, but not unlock, since you'd still have to clean the platform.

File: server/app.rb [hunk 3]
```diff
@@ -71,11 +83,17 @@ class App < Sinatra::Base
...
 
-    post '/unlock' do
+    delete '/lock' do
       require_basic_auth
-      File.delete(LOCK_FILE) if File.exist?(LOCK_FILE)
-      status 200
-      "Lock cleared!"
+      if File.exist?(LOCK_FILE)
+        File.delete(LOCK_FILE)
+        File.delete(PID_FILE) if File.exist?(PID_FILE)
```
Comment (ymendel) [inline server/app.rb]: What if the lock file doesn't exist and the pid file does?

**Comment by sshirokov** [comment]
I think you're right and I'm wrong. Carry on.

**Comment by ymendel** [comment]
I agree that the pid file should be for the `make` process and the lock file should be for the overall lock. I also like the idea of killing the process to cancel a print.

This looks good to me. I'm just not sure about that one edge case of maybe there being a pid file and no lock file.

File: server/app.rb (hunk 3 above)
Comment (skalnik) [inline server/app.rb]: Hmm… could this be the solution to our "Unlock before the printing is done" issue? Maybe we should only allow the unlock if the pidfile is clear?

**Comment by skalnik** [comment]
@ymendel What about now? As soon as the `make` exits, it should be cleared out, and we'll only allow folks to unlock if the pid file is missing.

**Comment by ymendel** [comment]
Yeah, that looks good.

**Pull request merged**